near-sdk/src/store/free_list/mod.rs

```
@@ -220,6 +220,86 @@ where
+    {
+        Defrag::new(self).defrag(callback);
+        self.first_free = None;
+    }
+}
+
+///Defrag struct has helper functions to perform defragmentation of `FreeList`. See the
+/// documentation of function `FreeList->defrag` for more details.
+struct Defrag<'a, T>
+where
+    T: BorshSerialize + BorshDeserialize,
+{
+    elements: &'a mut Vector<Slot<T>>,
+    occupied_count: u32,
+    curr_free_slot: Option<FreeListIndex>,
+    defrag_index: u32,
+}
+
+impl<'a, T> Defrag<'a, T>
+where
+    T: BorshSerialize + BorshDeserialize,
+{
+    /// Create a new struct for defragmenting `FreeList`.
+    fn new(list: &'a mut FreeList<T>) -> Self {
+        Self {
+            elements: &mut list.elements,
+            occupied_count: list.occupied_count,
+            defrag_index: list.occupied_count,
+            curr_free_slot: list.first_free,
+        }
+    }
+
+    fn defrag<F>(&mut self, mut callback: F)
+    where
+        F: FnMut(&T, u32),
+    {
+        while let Some(curr_free_index) = self.next_free_slot() {
+            if let Some((value, occupied_index)) = self.next_occupied() {
+                callback(value, curr_free_index.0);
+                //The entry at curr_free_index.0 should have `None` by now.
+                //Moving it to `occupied_index` will make that entry empty.
+                self.elements.swap(curr_free_index.0, occupied_index);
+            } else {
+                //Could not find an occupied slot to fill the free slot
+                env::panic_str(ERR_INCONSISTENT_STATE)
+            }
+        }
+        self.elements.len = self.occupied_count;
+    }
+
+    fn next_free_slot(&mut self) -> Option<FreeListIndex> {
+        while let Some(curr_free_index) = self.curr_free_slot {
+            let curr_slot = self.elements.values.remove(curr_free_index.0);
+            self.curr_free_slot = match curr_slot {
+                Some(Slot::Empty { next_free }) => next_free,
+                _ => None,
```

This branch seems like it should be unreachable, perhaps we want to panic with inconsistent state if hit, no? Or am I missing a case where this could happen?

---

It can have `None`  but not `Self::Occupied`. I'll fix it.